src/log/log.c

```diff
@@ -2918,12 +2918,20 @@ __log_write_internal(WT_SESSION_IMPL *session, WT_ITEM *record, WT_LSN *lsnp,
 	 */
 	if (ret != 0)
 		myslot.slot->slot_error = ret;
-	WT_ASSERT(session, ret == 0);
+	/*
+	 * If the addition of this record crosses the buffer boundary,
+	 * switch in a new slot. We have to do it after we release the
+	 * slot so that we cannot deadlock ourselves if all slots are in use.
+	 */
+	if (myslot.end_offset >= WT_LOG_SLOT_BUF_MAX ||
+	    F_ISSET(&myslot, WT_MYSLOT_UNBUFFERED) || force)
+		WT_TRET(__wt_log_slot_switch(
+		session, &myslot, true, false, NULL));
```

Comment:
Is this going in the opposite direction to https://jira.mongodb.org/browse/WT-4095 and the previous attempt in https://jira.mongodb.org/browse/WT-4058? I know that correctness definitely trumps performance, but I'd like to understand how these fit together. Will the ordering change here mean that we don't make a new slot available until after the I/O is performed?
Ah, I think the answer is "no" it's `__wt_log_release` that does a write so the switch still happens before that, right?

Reply:
Mostly the code movement here will cause joining threads to wait for the memory copy if a slot is full. So the thread that fills the slot buffer will do its copy into the memory buffer before setting up the new slot. The answer is mostly "no". The `log_fill` function in the common case will do the memory copy. The uncommon case involves I/O (write(2) to OS only) for unbuffered log records.